src/GitHub.VisualStudio/Services/VSGitExtFactory.cs

```diff
@@ -0,0 +1,49 @@
+﻿extern alias TF14;
+extern alias TF15;
+
+using System;
+using System.Threading.Tasks;
+using System.ComponentModel.Composition;
+using GitHub.Logging;
+using Serilog;
+using EnvDTE;
+using Microsoft.VisualStudio.Shell;
+using VSGitExt14 = TF14.GitHub.VisualStudio.Base.VSGitExt;
+using VSGitExt15 = TF15.GitHub.VisualStudio.Base.VSGitExt;
+
+namespace GitHub.Services
+{
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class VSGitExtFactory
+    {
+        static readonly ILogger log = LogManager.ForContext<VSGitExtFactory>();
+
+        readonly IGitHubServiceProvider serviceProvider;
+
+        [ImportingConstructor]
+        public VSGitExtFactory(IGitHubServiceProvider serviceProvider)
+        {
+            this.serviceProvider = serviceProvider;
+        }
+
+        [Export(typeof(IVSGitExt))]
+        public IVSGitExt VSGitExt => serviceProvider.GetService<IVSGitExt>();
+
+        public async static Task<IVSGitExt> Create(IAsyncServiceProvider sp)
+        {
+            var dte = await sp.GetServiceAsync(typeof(DTE)) as DTE;
+
+            // DTE.Version always ends with ".0" even for later minor versions.
+            switch (dte.Version)
+            {
+                case "14.0":
+                    return new Lazy<IVSGitExt>(() => new VSGitExt14(sp.GetServiceAsync)).Value;
```

Comment:
Why not just pass the `IAsyncServiceProvider` into `VSGitExt14/15` instead? 
Also, any specific reason why you need a `Lazy` here? You're evaluating it immediately after creating, which feels like it defeats the purpose of the Lazy in the first place.

Reply:
> Why not just pass the IAsyncServiceProvider into VSGitExt14/15 instead?

The multiple implementations of `IAsyncServiceProvider` wreck havoc. The `GitHub.TeamFoundation` assemblies pick up more dependencies and I end up having to use `extern alias`. Passing `GetServiceAsync` as a delegate ends up being much simpler.

> You're evaluating it immediately after creating, which feels like it defeats the purpose of the Lazy in the first place.

I've added some comments and change it to use `Func<IVSGitExt>` instead (using `Lazy` was a bit of a hack). By referencing `VSGitExt14` and `VSGitExt15` inside a lambda expression we can avoid loading the incompatible version into VS (e.g. we mustn't load `VSGitExt15` into VS 2015).